File: 92. Reverse Linked List II/92. Reverse Linked List II.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    if (!head || left === right) return head;
    
    const dummy = new ListNode(0);
    dummy.next = head;
    let prev: ListNode | null = dummy;
    
    for (let i = 0; i < left - 1; i++) {
        prev = prev!.next;
    }
    
    let start: ListNode | null = prev!.next;
    let then: ListNode | null = start!.next;
    
    for (let i = 0; i < right - left; i++) {
        start!.next = then!.next;
        then!.next = prev!.next;
        prev!.next = then;
        then = start!.next;
    }
    
    return dummy.next;
};

File: 92. Reverse Linked List II/92. Reverse Linked List II.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} left
# @param {Integer} right
# @return {ListNode}
def reverse_between(head, left, right)
    return head if head.nil? || left == right
    
    dummy = ListNode.new(0)
    dummy.next = head
    prev = dummy
    
    (left - 1).times do
        prev = prev.next
    end
    
    start = prev.next
    then_node = start.next
    
    (right - left).times do
        start.next = then_node.next
        then_node.next = prev.next
        prev.next = then_node
        then_node = start.next
    end
    
    dummy.next
end